### ISSUE TITLE
Infra: migrate to Truth in tests

### ISSUE THREAD

**Issue by pbludov** [checkstyle/checkstyle#9142]
We [agreed](https://github.com/checkstyle/checkstyle/issues/7239#issuecomment-560147839) to use [Truth](https://github.com/google/truth) in tests. It is more readable and produces better diffs on failure.

This is how a regular unit test should be written:
```java
    @Test
    public void testFomeMethod() {
        assertWithMessage("Human-readable assertion")
            .that(someClass.method(arg1, arg2))
            .isTrue()
            .contains(...)
            .isInstanceOf(...)
    }
```
Assertion starts with `assertWithMessage(...)`, followed by `that` and number of [assertions](https://truth.dev/benefits#readable-assertions).

For asserting exceptions:

```java
    @Test
    public void testMethodThrowsException() {
        final SomeExceptionType exception =  assertThrows(SomeExceptionType.class, () -> {
            someClass.someMethod(args);
        });
        assertWithMessage("Invalid exception message")
            .that(exception)
            .hasMessageThat()
                .isEqualTo("The expected message");
        assertWithMessage("Invalid inner exception")
            .that(exception)
            .hasCauseThat()
                .isInstanceOf(...)
                .hasMessageThat()
                    ...;
    }
```

**Comment by nmancus1** [comment]
@pbludov would truth be used for check unit tests? If so, would we deprecate usage of `verify()` and create [custom assertion methods](https://truth.dev/extension.html)?

**Comment by pbludov** [comment]
> @pbludov would truth be used for check unit tests? If so, would we deprecate usage of `verify()` and create [custom assertion methods](https://truth.dev/extension.html)?

That would be perfect. We should try to implement such extensions.

**Comment by nmancus1** [comment]
@pbludov since this issue is approved, should we begin using truth in all new unit tests going forward?

**Comment by romani** [comment]
Yes, we try to keep this in mind during review

**Comment by nmancus1** [comment]
> Yes, we try to keep this in mind during review

Should we create separate issue to https://github.com/checkstyle/checkstyle/issues/9142#issuecomment-758044801 so that we can use this in check UT's sooner?

**Comment by nmancus1** [comment]
Also, would this be a good GSOC 2nd issue? Break this issue into one for each test class?

**Comment by romani** [comment]
Yes, but we need to find test classes that need update, as majority of tests already abstracted a lot from core asserts, example https://github.com/checkstyle/checkstyle/commit/7190c47ca5515ad8cb827bc4065ae7664d2766c1

It would be good to think how we can enforce no new files with junit assert.

**Comment by strkkk** [comment]
What do you think - should we repeat code several times or extract method to make such things as one-liner?
I mean code like 
https://github.com/checkstyle/checkstyle/pull/10026/files#diff-cf838e5bbdb7de5a3e0e91f9919a389a0fd2ffdbda60c85a3c6dba07dcf69acdR365-R375
```
        assertWithMessage(message.toString())
                .that(missingViolations)
                .isEmpty();

        assertWithMessage(message.toString())
                .that(unexpectedViolations)
                .isEmpty();

        assertWithMessage(message.toString())
                .that(differingViolations)
                .isEmpty();
```
this can be extracted in one method and be like `assertIsEmpty(message, collection)`

What do you think?

**Comment by timurt** [comment]
@pbludov do we plan to fully migrate or some methods from junit will be allowed? for example I see that we allow usage of 
`import static org.junit.jupiter.api.Assertions.fail;`

**Comment by pbludov** [comment]
> @pbludov do we plan to fully migrate or some methods from junit will be allowed?

The goal of this issue is to make assertions (and the output) more readable. See this [comment](https://github.com/checkstyle/checkstyle/issues/7239#issuecomment-550014703).

We should remove all usages of AssertJ. The output of method `fail` from stock `JUnit` is clear enough, we can keep it. The `Trurh` framework has no method to replace it.

**Comment by rnveach** [comment]
Questions to resolve the last of this issue:

1) Is it valid to convert assertSame to isEqualTo for truth? Same ensure it is the same memory location, while I assume equal to ensures the "equal" method returns true.
2) What is the valid way to convert assertThrows to Truth? It uses a lambda and probably has an encapsulation for the try/catch.

**Comment by romani** [comment]
> Is it valid to convert assertSame to isEqualTo for truth?

https://truth.dev/api/latest/com/google/common/truth/Subject.html#isSameInstanceAs(java.lang.Object)
But most case that I see in our code, should be converted to isEquals (class type comparison).

> What is the valid way to convert assertThrows to Truth? It uses a lambda and probably has an encapsulation for the try/catch.

yes, we already do this.
https://github.com/checkstyle/checkstyle/blob/84eb861b9486022a273781046a3190c37604824d/src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java#L122-L128

**Comment by rnveach** [comment]
@romani 
> What is the valid way to convert assertThrows to Truth? It uses a lambda and probably has an encapsulation for the try/catch.
> yes, we already do this.

The code you quoted uses `assertThrows` which is what I am asking how to convert and remove. `assertThrows` is not a Truth assertion, and uses the forbidden library `org.junit.jupiter.api.Assertions` which we are trying to forbid and remove all uses of in this issue.

**Comment by rnveach** [comment]
> assertThrows 

Still waiting on a resolution for this.
This is the main thing left to handle for this issue.

**Comment by romani** [comment]
https://stackoverflow.com/a/38866403

No good API, let's keep assertThrows
